lib/ukboot: Shutdown feature

skuenzer opened the pull request:
### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Description of changes

This PR introduces support for shutting down the system. The change consists of an extension of inittab. Each entry is extended to two function pointers: an initialization function pointer and a termination function pointer, which are handled in pairs and share the same priority. While initialization means iterating over the inittab and calling the initialization functions, termination is done by iterating over the inittab in reverse order.
Initialization and termination are handled by the "init" thread of `lib/ukboot`. With scheduling, PR introduces the option to start the `main()` of the application in a separate "main" thread. This allows shutdown requests to be handled independently of the state of `main()`. A shutdown request driver (e.g., ACPI button) or any other application thread can send a request to the "init" thread to initiate a shutdown.

skuenzer committed:
include/uk: Registration of termination functions

This commit adopts the inittab registration macros so that termination
function can be handed over, too. This happens always in pair:
Inititalization function together with termination function.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
include/uk: Introduce context for init function

This commit introduces a context argument (`struct uk_init_ctx *`) to
the initialization functions. The motivation is that 1) the list of
arguments can be easily extended without changing the init function
signature and 2) we introduce a way to handover boot information
across initalization functions.
This commit adds the argument count (`argc`) and a reference to the
argument vector (`argv`) to `struct uk_init_ctx`. `lib/ukboot` is adopted
accordingly.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
include/uk: Extend inittab with termination functions

This commit extends the inittab entries with another function pointer:
a corresponding termination function. The functions are handled as pair
because the idea is that a shutdown means iterating over the table in
reverse order and calling the termination functions. `lib/ukboot` is
adopted accordingly.
Similar to the init functions, a context argument for the termination
functions is introduced, too: `struct uk_term_ctx *`. In its initial
version, it only contains the shutdown request (e.g., `UKPLAT_HALT`,
`UKPLAT_RESTART`).

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

Commit:
lib/ukboot: Provide argv for __init_array

Most UNIX-like operating systems call the constructors listed in
__init_array with the argument vector. This commit changes the
constructor function type to be non-prototyped so that we can
use the same type for all classes of constructors. The commit
then changes the boot library to supply the argument vector to
the __init_array constructors.

Signed-off-by: Marc Rittinghaus <marc@unikraft.io>

skuenzer committed:
lib/ukboot: Handle shutdown requests with "init" thread

This commit introduces a "main" thread for calling the application's
`main()` function. This is done so that "init" is purely responsible
for system initialization and shutdown. After creating the "main" thread,
"init" will block on a barrier which can only be unblocked by a
call to `uk_boot_shutdown_req()`. "init" will then perform a system
shutdown. With this model, any thread, including "main", can request
a shutdown.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/ukboot: Shutdown request event handler

This commit registers the shutdown request function
`uk_boot_shutdown_req_isr()` to the shutdown request event queue. This
allows drivers to trigger a shutdown which is then performed by the "init"
thread.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uklock: Provide ISR variant of `uk_semaphore_up()`

This commit provides an ISR-safe variant of `uk_semaphore_up()`. This
enables unblocking of threads that wait on a semaphore from an interrupt
context.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/ukboot: ISR variant of `uk_boot_shutdown_req()`

This commit provides an ISR-safe variant of `uk_boot_shutdown_req()`. THis
function is safe to be called from interrupt contexts, like a driver
interrupt. The commit tries to avoid to duplicate code and compiles
`shutdown_req.c` two times: once without ISR properties and another time
with ISR property.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uksched: Provide ISR variants for thread wake up interfaces

This commit provides ISR variants of the following thread wake up
functions:
 - `uk_waitq_wake_up()`
 - `uk_sched_thread_woken()`
 - `uk_thread_wake()`
This enables to call these functions from interrupt contextes. Due to
dependencies it is expected that an actual scheduler implementation is
providing an ISR variant of the `thread_woken_func` function.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/ukschedcoop: Implement ISR variant of `thread_woken_func`

This commit compiles the implementation of `schedcoop_thread_woken()`
as ISR-safe. The same function is used for both: non-ISR and ISR
contexts. This is safe to do because by definition an ISR-safe function
can be called from non-ISR context but not vice versa.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
include/uk: Notes about shutdown request event queue

This commit documents the shutdown request event queue within
`include/uk/plat/bootstrap.h`. The idea is that a platform declare the
queue and raises events with drivers. Any library can then register
shutdown handlers to this queue and can take according actions.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/uknetdev: Make `uk_netdev_drv_rx_event()` ISR-safe

This commit uses the ISR-safe semaphore variant (`uk_semaphore_up_isr()`)
so that `uk_netdev_drv_rx_event()` can be called from a drivers' interrupt
context.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

jschlumpp commented on include/uk/init.h:
```diff
@@ -55,88 +79,117 @@ typedef int (*uk_init_func_t)(void);
+		.term = (term_fn)					\
+	}
 
-#define _UK_INITTAB(fn, base, prio)					\
```
I would prefer to not change the existing interface and instead introduce a new macro that allows specifying the `term_fn`. Then it wouldn't break all existing (external) clients of the API.

skuenzer replied:
The update requires a macro that sets both at the same time because this is the way how the table is constructed, the same priority value is used for initialization and termination (reverse order). This simplifies cleaning up in case of errors of one of the init function.
Of course we could argue, another macro could be provided as wrapper that sets only init but I wanted to have this explicitly. The fact that the system can shutdown should also make developers think about a termination. As soon this is upstreamed, we have a number of PR across libs that fix existing code so that disruption should be minimal.

mogasergiu commented on lib/ukboot/boot.c:
```diff
@@ -419,38 +488,49 @@ void ukplat_entry(int argc, char *argv[])
+#if CONFIG_LIBUKBOOT_MAINTHREAD
+static __noreturn void main_thread(void *a_argc, void *a_argv)
+{
+	int argc    = (int)((__uptr)a_argc);
```
I know it looks nice, but I think we should maintain consistency from now on and have `=` separated by only one whitespeace end-to-end.

skuenzer replied:
Is fine, I don't mind. I will send an update.

mogasergiu commented on lib/uklock/include/uk/isr/semaphore.h:
```diff
@@ -1,8 +1,34 @@
+
+	uk_spin_lock_irqsave(&(s->sl), irqf);
+	++s->count;
+#ifdef UK_SEMAPHORE_DEBUG
```
why not `#if` (and closing `/* UK_SEMAPHORE_DEBUG */`).

skuenzer replied:
I'll update 👍🏻

mogasergiu commented on lib/ukboot/shutdown_req.c:
```diff
@@ -54,3 +55,17 @@ int uk_boot_shutdown_req_isr(enum ukplat_gstate target)
+		return UK_EVENT_NOT_HANDLED;
+	return UK_EVENT_HANDLED_CONT;
+}
+UK_EVENT_HANDLER(UKPLAT_SHUTDOWN_EVENT, shutdown_req_handler);
```
One question. What happens if there are multiple shutdown handlers registered? Are they executed based on a priority, like on Linux? E.g. UEFI shutdown routine > ACPI shutdown routine

skuenzer replied:
UEFI and ACPI would be triggers of the shutdown event queue. Whatever pulls the trigger first, made it actually.

mogasergiu replied:
UEFI shutdown services are at the top of the hierarchy in the list of shutdown execution. An ACPI power button event may trigger it but even then, it's still the UEFI shutdown that should be done to shutdown the machine. If that fails or is somehow absent, then yes fallback to ACPI shutdown.

mogasergiu reviewed (commented):
Hi Simon :wave:! Had a quick look, very clean PR! I like the semaphore idea! The only downside of this PR I can see is the shutdown event not being handled with highest priority (i.e. it has to wait to be scheduled in the round-robin fashion) which can be risky in the case of high emergency power management events, especially in a cooperative scheduler environment (maybe something like a deferred handler could solve this :pleading_face:)

Anyway, I left some minor comments. I will probably have another look in the near future :cowboy_hat_face:.

skuenzer commented:
> Hi Simon 👋! Had a quick look, very clean PR! I like the semaphore idea! The only downside of this PR I can see is the shutdown event not being handled with highest priority (i.e. it has to wait to be scheduled in the round-robin fashion) which can be risky in the case of high emergency power management events, especially in a cooperative scheduler environment (maybe something like a deferred handler could solve this 🥺)
> 
> Anyway, I left some minor comments. I will probably have another look in the near future 🤠.

Hi Sergiu, thanks a lot for your review. It is in fact true that a shutdown request can be ignored as long as a thread never releases the CPU which lets the init thread not being scheduled to forward a shutdown request to the application. However, I think this behavior is fine since we are speaking here about a shutdown request which the application could even ignore. So, my assumption is that this is a cooperative action. With preempt scheduling this might be even less of a problem and the hypervisor can still terminate non-cooperating virtual machines.

mogasergiu commented:
> can be ignored as long as a thread never releases the CPU

Yes you are right. I was however referring to the situation where this is not the next thread that will be executed on yield, because we do not have a priority scheduling schema, right?

>  hypervisor can still terminate non-cooperating virtual machines.

Yes, agreed, not something worth bothering with now :smiley:. I just thought it would be worth a mention :wink:.

mogasergiu commented on lib/ukboot/boot.c:
```diff
@@ -339,6 +341,9 @@ void ukplat_entry(int argc, char *argv[])
 	uk_sched_start(s);
 #endif /* !CONFIG_LIBUKBOOT_NOSCHED */
 
+	ictx.cmdline.argc = argc - kern_args;
```
Hmm, isn't `kern_args` always `0`? Ditto for line below.

skuenzer replied:
Ouh, this seems to come from a leftover before we updated `libuklibparam`. I can fix it with this commit. 👍🏻

mogasergiu commented on include/uk/init.h:
```diff
@@ -126,26 +144,46 @@ typedef int (*uk_init_func_t)(struct uk_init_ctx *);
+ */
+#define uk_inittab_foreach_reverse(itr, inittab_start, inittab_end)	\
+	for ((itr) = (DECONST(struct uk_inittab_entry *,		\
+			 (&(inittab_end)))) - 1);			\
```
Perhaps we could align this to the `DECONST` open paranthesis above?

skuenzer replied:
Sure 👍🏻

mogasergiu commented on include/uk/ctors.h:
```diff
@@ -42,7 +42,7 @@
 #endif
 
-typedef void (*uk_ctor_func_t)(void);
+typedef void (*uk_ctor_func_t)();
```
Why no arguments? Shouldn't it be the types of `argc`/`argv`?

skuenzer replied:
To make it compatible with any number of function arguments even though nothing is handed over.

mogasergiu commented on lib/ukboot/boot.c:
```diff
@@ -377,12 +392,65 @@ void ukplat_entry(int argc, char *argv[])
+static inline int do_main(int argc, char *argv[])
+{
+#if CONFIG_LIBPOSIX_ENVIRON
+	char **envp;
```
Aw man, it was almost an xmas tree style local variable declaration :cry:, maybe we can make it like that and slowly turn the codebase to follow that aesthetic pattern :pray:.

skuenzer replied:
Since when are we following christmas tree order of variable declarations, I usually group them into logical connection and then apply trees per group... I can update this one just for you 😄

mogasergiu commented on lib/ukboot/include/uk/boot.h:
```diff
@@ -0,0 +1,44 @@
+}
+#endif
+#endif /* !__ASSEMBLY__ */
+#endif /* !__UK_BOOT_BOOT_H__ */
```
Why `!__UK_BOOT_BOOT_H__`?

skuenzer replied:
Because of libukboot and boot.h:
`__UK_LIBNAME_HEADERNAME_H__`

mogasergiu replied:
Yes I meant why the `!`?

skuenzer replied:
:magic_wand: You did not see this 😄

mogasergiu commented on lib/uksched/include/uk/isr/sched.h:
```diff
@@ -0,0 +1,25 @@
+/* ISR variant of `uk_sched_thread_woken()` */
+static inline void uk_sched_thread_woken_isr(struct uk_thread *t)
+{
+
```
Redundant blank line?

skuenzer replied:
Yes 👍🏻

mogasergiu commented on lib/uksched/include/uk/isr/wait.h:
```diff
@@ -0,0 +1,23 @@
+	struct uk_waitq_entry *curr, *tmp;
+	unsigned long flags;
+
+	ukplat_spin_lock_irqsave(&(wq->sl), flags);
```
Are paranthesis required around `wq->sl`?

skuenzer replied:
Probably not, the precedence of C operators does not require it. Sometimes I am paranoid and want to be sure that the order is correct. I'll remove it. 👍🏻

mogasergiu commented on lib/uksched/include/uk/sched_impl.h:
```diff
@@ -55,15 +55,17 @@ int uk_sched_register(struct uk_sched *s);
-		(s)->thread_woken    = thread_woken_func; \
-		(s)->idle_thread     = idle_thread_func; \
+		(s)->sched_start      = start_func; \
+		(s)->yield            = yield_func; \
```
Could we align these backslashes?

skuenzer replied:
This would require aligning all backslahses of this Macro defintion. I am never a fan of too intrusive alignment changes so I adopted to what was there already.

mogasergiu commented on lib/ukboot/include/uk/isr/boot.h:
```diff
@@ -0,0 +1,45 @@
+}
+#endif
+#endif /* !__ASSEMBLY__ */
+#endif /* !__UK_BOOT_BOOT_H__ */
```
`__UK_BOOT_BOOT_ISR_H__`?

skuenzer replied:
uups, yes 👍🏻

mogasergiu commented on lib/ukboot/shutdown_req.c:
```diff
@@ -54,3 +55,17 @@ int uk_boot_shutdown_req_isr(enum ukplat_gstate target)
+		return UK_EVENT_NOT_HANDLED;
+	return UK_EVENT_HANDLED_CONT;
+}
+UK_EVENT_HANDLER(UKPLAT_SHUTDOWN_EVENT, shutdown_req_handler);
```
`checkpatch.pl --strict` throws
```
CHECK: Please use a blank line after function/struct/union/enum declarations
#62: FILE: lib/ukboot/shutdown_req.c:70:
+}
+UK_EVENT_HANDLER(UKPLAT_SHUTDOWN_EVENT, shutdown_req_handler);
```

skuenzer replied:
This is a registration call and not a (function prototype) declaration. I think checkpatch is too strict here, however, I can add a line in between.

mogasergiu reviewed (changes requested):
Ok had my last look. After this I should be able to give my tag. Everything still looks very good :raised_hands:, mostly just some (perhaps subjective?) styling comments.

StefanJum committed:
lib/uklock: Fix recursive mutex static inititializer

The `struct uk_mutex` structure expects the members to be, in order,
`lock_count`, `flags`, `owner`, `waitq`. Change the static inititializer
to set the flags to `UK_MUTEX_CONFIG_RECURSE` instead of the owner.

Signed-off-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>
Approved-by: Marco Schlumpp <marco@unikraft.io>
GitHub-Fixes: #1195

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
